### PR TITLE
Add failing test case with <select> and propose a fix

### PR DESCRIPTION
**Pull request by soyuka**
Hi, first great job with choo and all the `nanomodules` love them <3.

I think there's an issue on diffing `<select><option selected="selected">value</option></select>`. Indeed, the first diff is always correct, but it fails the second time. It gets worse when you add more than 2 `<option>` elements.

Note that I used `selected="${condition ? 'selected' : null}"` as stated in the docs, `null` should remove the attribute? 

Thanks!

### CODE REVIEW

**Comment by soyuka** [comment]
I've submitted a patch for the failing test case but I'm not sure it's correct. I think that the `selected` attribute should first be removed when it's value is not `"selected"`. Because of how bel works, `null` actually evaluates to `'null'`, which is semantically incorrect IMO. 

Also, because of how `selectedIndex` is computed (ie `node.value === node.options[node.selectedIndex].value`) it might be best to first diff `OPTION` (change selected attribute), which will give the correct new index.

Lmk if you have a better suggestion on how this can be fixed! Thanks!

File: lib/morph.js [hunk 1]
```diff
@@ -146,18 +146,20 @@ function updateTextarea (newNode, oldNode) {
...
-        if (curChild.hasAttributeNS(null, 'selected')) break
+        if (curChild.hasAttributeNS(null, 'selected') && curChild.getAttributeNS(null, 'selected') !== 'null') break
         i++
       }
+
       curChild = curChild.nextSibling
     }
 
-    newNode.selectedIndex = i
+    newNode.selectedIndex = oldNode.selectedIndex = i
```
Comment (yoshuawuyts) [inline lib/morph.js]: This should suffice:
```js
oldNode.selectedIndex = i
```

Comment (soyuka) [inline lib/morph.js]: doesn't :|

Comment (soyuka) [inline lib/morph.js]: In fact, if we don't update the newNode selectedIndex here, it keeps the value of the previous node (not sure why). 

For example in the test case:

```javascript
      var values = ['foo', 'bar']

      var a = html`<select>${values.map(function (e) { return html`<option value="${e}" selected="${e === 'bar' ? 'selected' : null}">${e}</option>` })}</select>`
      var b = html`<select>${values.map(function (e) { return html`<option value="${e}" selected="${e === 'foo' ? 'selected' : null}">${e}</option>` })}</select>`

      var res = nanomorph(a, b)
      t.equal(res.value, 'foo')
```

b.selectedIndex is `1`, but we choose `0`. If we don't update it, it stays that way and the resulting morph is wron. I do agree though that this should not be necessary. I think I've not enough knowledge about what goes down before we reach the `updateSelect` part to find out why...

**Comment by yoshuawuyts** [comment]
FYI if you use `false` instead of `null` it should work as expected. Been trying to get `null` to work properly and that one's slightly trickier.

**Comment by soyuka** [comment]
Haha nice one, should I add this to the readme in another PR? 

the `oldNode.selectedIndex = i` is indeed fixing the `false` issue.

File: lib/morph.js [hunk 2]
```diff
@@ -157,7 +157,7 @@ function updateSelect (newNode, oldNode) {
       curChild = curChild.nextSibling
     }
 
-    newNode.selectedIndex = i
+    oldNode.selectedIndex = i
```
Comment (soyuka) [inline lib/morph.js]: this is still wrong, don't merge

**Comment by soyuka** [comment]
So, after some more digging I'm not able to use a condition in the `selected` attribute IMO there are too many edge cases and I think that the `selected` should only be there once. 

Anyway, as we're playing with real dom here (<3) I'm using:

```
var option = html`<option value="foo">Foo</option>`
if (mySelectedCondition === true) {
  option.setAttribute('selected', 'selected')
}
```

This patch is still needed to update `oldNode` instead of `newNode`.

ping @louiscenter

**Comment by tornqvist** [comment]
I just opened another pull request (#56) before I noticed this one. It's dealing with the same issue only that I think this test case is testing two different things, there's `selected=${ null }` and the actual updating of the selected value.

If you were to hard code the test html (or use `false` in place of `null`) you'll see that this fix does work (see aforementioned PR).

However, this fix does not handle the case where new options are added to the select element, since it's walking the old tree and not the new one. The following change solves that and is in my other pull request. I'm not sure if it'd be better to re-open that or to apply the same fix here.

```diff
- var curChild = oldNode.firstChild
+ var curChild = newNode.firstChild
```

**Comment by soyuka** [comment]
Thanks @tornqvist ! I've added a test case while I was at it.

**Comment by Flet** [comment]
Applying this change manually makes my `selected` options work beautifully. Please merge 🙏

**Comment by yoshuawuyts** [comment]
Closing in favor of https://github.com/yoshuawuyts/nanomorph/pull/60 - includes same tests, but introduces a slightly different fix

**Comment by yoshuawuyts** [comment]
`v5.0.0`

**Pull request closed without merging**